### PR TITLE
[TRTLLM-12982][chore] relocate `torch_multi_arange`

### PR DESCRIPTION
**Pull request by ixlmar**
## Description

Follow-up on https://github.com/NVIDIA/TensorRT-LLM/pull/14693#discussion_r3351423332.

Commit 800c7ee40ffa7680ecfb3f2662da427ead34131c is from #15413, which is to be merged before this PR.

## Test Coverage

Covered by existing tests

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Summary by CodeRabbit

* **Improvements**
  * Encoder CUDA graphs now properly detect multi-item scoring scenarios and fall back to eager execution when necessary.

* **Refactoring**
  * Optimized attention metadata to accept multi-item configuration during preparation phase instead of forward pass.
  * Reorganized utility functions for improved code maintainability.

* **Chores**
  * Updated test infrastructure and file organization.

### CODE REVIEW

**Review by Funatiq — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]

**Review by brb-nv — APPROVED** [review]
LGTM.

File: legacy-files.txt [hunk 1]
```diff
@@ -1004,7 +1004,7 @@ tests/unittest/_torch/ray_orchestrator/multi_gpu/test_ops.py
 tests/unittest/_torch/ray_orchestrator/single_gpu/test_cache_transceiver_comm.py
 tests/unittest/_torch/sampler/test_beam_search.py
 tests/unittest/_torch/sampler/test_best_of_n.py
-tests/unittest/_torch/sampler/test_torch_multi_arange.py
+tests/unittest/_torch/test_torch_multi_arange.py
```
Comment (yuxianq) [inline legacy-files.txt]: Use alphabetical order for path

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 2]
```diff
@@ -308,7 +308,7 @@ def num_ctx_tokens(self) -> int:
     def num_tokens(self) -> int:
         return self._num_tokens
 
-    def prepare(self):
+    def prepare(self, *, multi_item_part_lens: list[list[int]] | None = None):
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/interface.py]: Don't change the interface of `prepare`, if you need to provide `multi_item_part_lens` for `prepare`, add a field `multi_item_part_lens` to metadata, set it before `prepare` and access it in `prepare`.

**Review by MartinMarciniszyn — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]
Approved

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 2 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/interface.py]: Addressed in #15413

File: legacy-files.txt (hunk 1 above)
Comment (ixlmar) [reply legacy-files.txt]: Done

**Review by yuxianq — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 3]
```diff
@@ -722,6 +726,89 @@ def _plan_ragged_no_kv(
...
+            pin_memory=prefer_pinned(),
+            dtype=torch.uint32,
+        ).to(device=device, non_blocking=True)
+        token_pos_in_items_raw_lens = [  # 'raw' lengths before padding
+            sum(req_part_lens[1:]) + len(req_part_lens)
+            for req_part_lens in multi_item_part_lens
+        ]
+        token_pos_in_items_len = max(token_pos_in_items_raw_lens)
+        max_item_len_ptr = torch.tensor(
+            [max(req_part_lens[1:]) for req_part_lens in multi_item_part_lens],
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:761]: Out of scope for this PR

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 4]
```diff
@@ -722,6 +726,89 @@ def _plan_ragged_no_kv(
...
+        # every request, followed by [0] (final delimiter) which is fused with padding for simplicity.
+        range_ends = torch.tensor(
+            [
+                item_len + 1
+                for req_part_lens, token_pos_in_items_raw_len in zip(
+                    multi_item_part_lens, token_pos_in_items_raw_lens)
+                for item_len in (
+                    req_part_lens[1:] +
+                    [token_pos_in_items_len - token_pos_in_items_raw_len])
+            ],
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:776]: Out of scope for this PR

File: tensorrt_llm/llmapi/llm.py [hunk 5]
```diff
@@ -899,6 +901,36 @@ def encode(
...
+                ],
+                pin_memory=prefer_pinned(),
+                dtype=torch.int32,
+            ).to("cuda", non_blocking=True)
+            position_ids_cuda = torch_multi_arange(
+                starts=starts_cuda,
+                ends=ends_cuda,
+                output_length=len(flat_token_ids),
+            )
+            forward_inputs["position_ids"] = position_ids_cuda
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/llmapi/llm.py]: Out of scope for this PR

File: tensorrt_llm/_torch/utils.py [hunk 6]
```diff
@@ -520,3 +520,96 @@ def replace_parameter_and_save_metadata(
...
+    zeros = torch.zeros((), dtype=ends.dtype, device=ends.device)
+    if steps is None:
+        steps = ones.broadcast_to(ends.shape)
+    jumps = -prev_range_ends  # delta from one range to the next
+    if starts is not None:
+        jumps += starts
+    #     NB: Apply correction for empty ranges
+    jumps_corrections = torch.where(repeats == 0, jumps,
+                                    zeros).cumsum(0, dtype=ends.dtype)
+    jumps += jumps_corrections
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/utils.py]: Done

File: tensorrt_llm/_torch/utils.py [hunk 7]
```diff
@@ -520,3 +520,96 @@ def replace_parameter_and_save_metadata(
...
+    #    preceding range into the start element for the current range and 'b' is
+    #    simply the step size for the current range.
+    #
+    repeats = ends  # number of elements in each range
+    if starts is not None:
+        repeats = repeats.clone()
+        repeats -= starts
+    if steps is not None:
+        repeats *= steps.sign()
+        steps_abs = steps.abs()
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/utils.py]: Done

File: tensorrt_llm/_torch/utils.py [hunk 8]
```diff
@@ -520,3 +520,96 @@ def replace_parameter_and_save_metadata(
...
+    or when tensors are known to reside on the host.
+    """
+    if steps is not None:
+        assert ends.dtype == steps.dtype
+        assert ends.shape == steps.shape
+        assert ends.device == steps.device
+    if starts is not None:
+        assert ends.dtype == starts.dtype
+        assert ends.shape == starts.shape
+        assert ends.device == starts.device
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/utils.py]: Done

File: tensorrt_llm/llmapi/llm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/llmapi/llm.py]: Addressed in #15627

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:776]: Addressed in #15627

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:761]: Addressed in #15627
